### PR TITLE
[develop] browser: Add support for browsing and extracting zip files

### PR DESCRIPTION
**Pull request by VicesOfTheMind**
## Description
This PR allows the browser to open ZIP files as directories and extract files from the archive. For simplicity ZIP archives are treated as flat directories with each entry simply showing the full folder hierarchy. However the appropriate folder structure is still created when extracting a file from a subfolder in the ZIP.

## Motivation and Context
Sometimes you might put a zipped ROM on your SD card when not paying attention. Rather than having to put the SD card back into your PC you can just extract it using your N64 instead.

## How Has This Been Tested?
I've tested it by extracting various files from `sc64-extra-v2.20.2.zip` to test extracting files with subfolders.
I've also created my own test zip to test empty folders and file comments.

## Screenshots
screenshot

screenshot

## Types of changes
- [x] Improvement (non-breaking change that adds a new feature)
- [ ] Bug fix (fixes an issue)
- [ ] Breaking change (breaking change)
- [x] Documentation Improvement
- [x] Config and build (change in the configuration and build system, has no impact on code or features)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

You agree with the license terms and that other license types may be granted with permission of the original `N64FlashcartMenu` project license holders.

## Summary by CodeRabbit

- New Features
  - Browse ZIP archives in the file browser, view contents, and extract individual files with real-time progress and clear error messages.
  - Archive-aware navigation and context actions (open archive, show entry properties, extract).
  - Detailed entry information screen (size, compression, directory/file status, encryption, comments).
- Style
  - Distinct styling for archive items in file lists to improve recognition.

### CODE REVIEW

File: src/menu/menu_state.h [hunk 1]
```diff
@@ -119,6 +128,7 @@ typedef struct {
         bool rom_file;
         bool disk_file;
         bool emulator_file;
+        bool extract_file;
     } boot_pending;
```
Comment (VicesOfTheMind) [inline src/menu/menu_state.h]: Is it alright if I rename this to `load_pending`?

File: src/menu/menu_state.h [hunk 2]
```diff
@@ -56,14 +59,17 @@ typedef enum {
...
+    ENTRY_TYPE_ARCHIVE,
+    ENTRY_TYPE_ARCHIVED
 } entry_type_t;
 
 /** @brief File Entry Structure */
 typedef struct {
     char *name;
     entry_type_t type;
     int64_t size;
+    int index;
```
Comment (networkfusion) [inline src/menu/menu_state.h]: use int types e.g. int32_t

File: src/menu/views/browser.c [hunk 3]
```diff
@@ -293,8 +352,34 @@ static bool pop_directory (menu_t *menu) {
...
+    menu->browser.directory = path_clone(file);
+    path_pop(menu->browser.directory);
+
+    if (load_directory(menu)) {
+        path_free(menu->browser.directory);
+        menu->browser.directory = previous_directory;
+        return true;
+    }
+
+    for (unsigned short i = 0; i < menu->browser.entries; i++) {
```
Comment (networkfusion) [inline src/menu/views/browser.c]: use int types e.g. uint16_t

File: src/menu/views/browser.c [hunk 4]
```diff
@@ -150,6 +162,50 @@ static void browser_list_free (menu_t *menu) {
...
+
+    if (menu->browser.entries > 0) {
+        menu->browser.selected = 0;
+        menu->browser.entry = &menu->browser.list[menu->browser.selected];
+    }
+
+    qsort(menu->browser.list, menu->browser.entries, sizeof(entry_t), compare_entry);
+
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (VicesOfTheMind) [reply src/menu/views/browser.c:209]: This is a fair assessment, but since stat can only fail due to invalid arguments I simply changed the code to abort opening the zip archive rather than deal with `realloc`.

File: src/menu/views/extract_file.c [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            menu->browser.select_file = path_clone(path);
+            menu->next_mode = MENU_MODE_BROWSER;
+        } else {
+            menu_show_error(menu, "Failed to create file");
+        }
+    }
+
+    path_free(path);
+    path_free(dir);
+}
```
[the remark this answers is not in the document]
Comment (VicesOfTheMind) [reply src/menu/views/extract_file.c:61]: The current code will not overwrite existing files, so extracting into an unintended directory isn't much of a problem.

File: src/menu/views/extract_file.c [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        return " Type: Music file\n";
+    } else if (file_has_extensions(name, controller_pak_extensions)) {
+        return " Type: Controller Pak file\n";
+    } else if (file_has_extensions(name, emulator_extensions)) {
+        return " Type: Emulator ROM file\n";
+    } else if (file_has_extensions(name, cheat_extensions)) {
+        return " Type: Cheats file\n";
+    }
+    return " Type: Unknown file\n";
+}
```
Comment (VicesOfTheMind) [inline src/menu/views/extract_file.c]: All of the above code has been copied from the other file info view. I could move this into the `fs` utility module to avoid code duplication.

File: src/menu/menu_state.h (hunk 1 above)
Comment (networkfusion) [reply src/menu/menu_state.h]: perhaps something like `archive_extraction_pending`... But why would it be in the `boot_pending` struct, as this is for what is about to occur on the flashcart (to boot the ROMs)

File: src/menu/views/extract_file.c (hunk 6 above)
Comment (networkfusion) [reply src/menu/views/extract_file.c]: I think it is possibily more than that (if needing to handle all the same types as the browser) and needs to be moved to a seperate `file_extension_handler.h`

File: src/menu/menu_state.h (hunk 1 above)
Comment (VicesOfTheMind) [reply src/menu/menu_state.h]: That's not how it's being used in practice. In practice these booleans just allows each view to track that something needs to be loaded and that the progress bar needs to be shown rather than the regular UI. It is not an accurate signal of what is about to occur on the flashcart. For example for ROM loading the boolean is immediately set to false while the boot is actually still pending:

```
    if (menu->boot_pending.disk_file) {
        menu->boot_pending.disk_file = false;
        load(menu);
    }
```

Calling it `load_pending` is a more accurate reflection of what those booleans actually mean.

File: src/menu/views/extract_file.c (hunk 6 above)
Comment (VicesOfTheMind) [reply src/menu/views/extract_file.c]: In the end I resolved this by adding a new `file_info` UI component instead, that way even more UI code can be centralized in one place.

File: src/menu/menu_state.h (hunk 1 above)
Comment (networkfusion) [reply src/menu/menu_state.h]: @VicesOfTheMind Yes, after looking closer, I agree. Feel free to rename it.

**Comment by networkfusion** [comment]
Note: https://github.com/richgel999/miniz/pull/359

**Comment by VicesOfTheMind** [comment]
@networkfusion Made the change to `boot_pending`, changes to `utime` in either libdragon or miniz are needed to allow displaying the modification time for zipped files and to accurately reflect the modification time on extracted files.

**Comment by networkfusion** [comment]
> @networkfusion Made the change to `boot_pending`, changes to `utime` in either libdragon or miniz are needed to allow displaying the modification time for zipped files and to accurately reflect the modification time on extracted files. However that is also just perfectionism on my part, we could merge the PR without it and add it in later.

Great, and thanks. I will merge this now, and hope for your perfection PR 😜
